libpod/runtime_volume_linux.go

```diff
@@ -48,6 +48,15 @@ func (r *Runtime) newVolume(ctx context.Context, options ...VolumeCreateOption)
 	}
 	volume.config.CreatedTime = time.Now()
 
+	// Check if volume with given name exists.
+	exists, err := r.state.HasVolume(volume.config.Name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error checking if volume with name %s exists", volume.config.Name)
+	}
+	if exists {
+		return nil, errors.Wrapf(define.ErrVolumeExists, "volume with name %s already exists", volume.config.Name)
+	}
+
```

Comment:
Just thinking to narrow the small race.  Would it make sense to move the next if statement `if volume.config.Driver == define.VolumeDriverLocal {` above these new lines, then move these new lines and the code up to the line `lock, err := r.lockManager.AllocateLock()` call at line 93 to a new function.  Then you could lock up that function and do a defer unlock for when it exited?  Real rough:

```
checkVolumeName(volume) volume {
lock
defer unlock 

	// Check if volume with given name exists.
	exists, err := r.state.HasVolume(volume.config.Name)
	if err != nil {
		return nil, errors.Wrapf(err, "error checking if volume with name %s exists", volume.config.Name)
	}
	if exists {
		return nil, errors.Wrapf(define.ErrVolumeExists, "volume with name %s already exists", volume.config.Name)
	}

// Create the mountpoint of this volume
	volPathRoot := filepath.Join(r.config.VolumePath, volume.config.Name)
	if err := os.MkdirAll(volPathRoot, 0700); err != nil {
		return nil, errors.Wrapf(err, "error creating volume directory %q", volPathRoot)
	}
	if err := os.Chown(volPathRoot, volume.config.UID, volume.config.GID); err != nil {
		return nil, errors.Wrapf(err, "error chowning volume directory %q to %d:%d", volPathRoot, volume.config.UID, volume.config.GID)
	}
	fullVolPath := filepath.Join(volPathRoot, "_data")
	if err := os.Mkdir(fullVolPath, 0755); err != nil {
		return nil, errors.Wrapf(err, "error creating volume directory %q", fullVolPath)
	}
	if err := os.Chown(fullVolPath, volume.config.UID, volume.config.GID); err != nil {
		return nil, errors.Wrapf(err, "error chowning volume directory %q to %d:%d", fullVolPath, volume.config.UID, volume.config.GID)
	}
	if err := LabelVolumePath(fullVolPath, true); err != nil {
		return nil, err
	}
	volume.config.MountPoint = fullVolPath

return volume, nil
}
```

Reply:
The problem is that we don't have a global runtime lock right now, so we can't effectively lock this portion